### ISSUE TITLE
cergyk - BBLiquidation::_liquidateUser liquidator can bypass protocol fee on liquidation by returning returnedShare == borrowShare

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-02-tapioca-judging#33]
cergyk

high

# BBLiquidation::_liquidateUser liquidator can bypass protocol fee on liquidation by returning returnedShare == borrowShare

## Summary
When a liquidator liquidates a position on BigBang/Singularity market, they do not get the full liquidationBonus amount, but a callerShare which depends on the efficiency of the liquidation. 

However since this share is taken after the liquidator has swapped the seized collateral to an asset amount, the liquidator can simply choose to return enough asset to repay the borrow, reducing the extra amount to zero.

In that case the protocol fee and the caller share would be zero, but the liquidator has seized the full liquidation bonus during the swap.

## Vulnerability Detail

We can see that after the collateral has been seized from the liquidatee, the full amount of collateral with the liquidation bonus is sent to an arbitrary `liquidationReceiver` during `_swapCollateralWithAsset`:

https://github.com/sherlock-audit/2024-02-tapioca/blob/main/Tapioca-bar/contracts/markets/bigBang/BBLiquidation.sol#L262-L263

https://github.com/sherlock-audit/2024-02-tapioca/blob/main/Tapioca-bar/contracts/markets/bigBang/BBLiquidation.sol#L152-L159

The liquidator can choose to send back only `borrowAmount` of asset, effectively keeping excess collateral to himself

## Impact
Liquidator steals the due fee from protocol during liquidation 

## Tool used

Manual Review

## Recommendation
Consider adding a slippage control to the swap executed by the liquidator (e.g the liquidator must return at least 105% of `borrowAmount`, when seizing 110% of equivalent collateral)

**Comment by cryptotechmaker** [comment]
Medium.
The issue seems to be valid. However the user is not able to steal the full collateral, but only the bonus part, out of which he would have taken 90% anyway.

**Comment by CergyK** [comment]
Escalate

This should be of high severity, as demonstrated, it enables to bypass protocol fees on any liquidation.
For reference, issue #148 which claims freezing of the same protocol fees is rated as high

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This should be of high severity, as demonstrated, it enables to bypass protocol fees on any liquidation.
> For reference, issue #148 which claims freezing of the same protocol fees is rated as high

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
@CergyK Fair enough, the impact is similar to #148 so could be high severity (albeit protocol fees are not locked, just completely lost)

**Comment by cvetanovv** [comment]
I agree with the escalation. The impact is the same as #148.

**Comment by cvetanovv** [comment]
Planning to accept the escalation and make this issue a valid High.

**Comment by Evert0x** [comment]
Result:
High
Has Duplicates

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [CergyK](https://github.com/sherlock-audit/2024-02-tapioca-judging/issues/33/#issuecomment-2031295607): accepted
